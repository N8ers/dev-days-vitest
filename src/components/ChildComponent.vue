<template>
  <div class="container">
    <h3>Child</h3>
    <p>{{ props.message }}</p>
    <hr />
    <input v-model="childMessage" type="text" />
    <button @click="emitToParent">emit message to parent!</button>
  </div>
</template>

<script setup>
import { ref } from "vue";

const props = defineProps({
  message: {
    type: String,
    required: false,
    default: "I am the default message",
  },
});

const emit = defineEmits(["emitMessageUp"]);

const childMessage = ref("");

const emitToParent = () => {
  emit("emitMessageUp", childMessage.value);
};
</script>

<style scoped>
.container {
  margin: 20px;
  border: 1px solid black;
  padding: 10px;
  width: 50%;
}
</style>
