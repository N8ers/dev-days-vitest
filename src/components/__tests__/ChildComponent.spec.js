import { mount } from "@vue/test-utils";
import { describe, expect, test } from "vitest";

import ChildComponent from "@/components/ChildComponent.vue";

describe("ChildComponent.vue", () => {
  describe("Renders content on load", () => {
    test("Fresh page snapshot", () => {
      // NATHAN - go through and make sure all the snaps make sense and it's how vitest recomends
      // const wrapper = mount(ChildComponent)
      // expect(wrapper).toMatchSnapshot()
    });
  });

  describe("Component handles props", () => {
    test("Component handles default 'message' prop", () => {
      const wrapper = mount(ChildComponent);

      expect(wrapper.vm.message).toBe("I am the default message");

      const pTag = wrapper.find("p");
      expect(pTag.text()).toBe("I am the default message");
    });

    test("Component handles defined 'message' prop", () => {
      const wrapper = mount(ChildComponent, {
        props: {
          message: "I am a message from the parent!",
        },
      });

      expect(wrapper.vm.message).toBe("I am a message from the parent!");

      const pTag = wrapper.find("p");
      expect(pTag.text()).toBe("I am a message from the parent!");
    });
  });

  describe("Component handles emits", () => {
    test("component handles 'emitMessageUp' emit", async () => {
      const wrapper = mount(ChildComponent);

      // Find input
      const textInput = wrapper.find("input");

      // Simulating typing into the input
      textInput.element.value = "A nice message :)";
      textInput.trigger("input");

      // click the button to emit - you can find and click something all at once
      await wrapper.find("button").trigger("click");

      // assert that emit fired
      expect(wrapper.emitted().emitMessageUp).toBeTruthy();

      // assert number of times the emit occured
      expect(wrapper.emitted().emitMessageUp.length).toBe(1);

      // assert the value the the emit sent
      expect(wrapper.emitted().emitMessageUp[0]).toEqual(["A nice message :)"]);

      // you can get a list of emits that happened on the component with `emittedByOrder()`
      // const orderedEmitEvents = wrapper.emittedByOrder();
    });
  });
});
