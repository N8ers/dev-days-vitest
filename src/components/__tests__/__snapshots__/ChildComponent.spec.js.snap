// Vitest Snapshot v1

exports[`ChildComponent.vue > Renders content on load > Fresh page snapshot 1`] = `
<div class="container" data-v-71c74cf1="">
  <h3 data-v-71c74cf1="">Child</h3>
  <p data-v-71c74cf1="">I am the default message</p>
  <hr data-v-71c74cf1=""><input type="text" data-v-71c74cf1=""><button data-v-71c74cf1="">emit message to parent!</button>
</div>
`;
